clang/runtime/dpct-rt/include/dpl_extras/vector.h

```diff
@@ -66,143 +67,206 @@ class device_vector {
 
   void _set_capacity_and_alloc() {
     _capacity = ::std::max(_size * 2, _min_capacity());
-    _storage = _alloc.allocate(_capacity);
+    _storage = ::std::allocator_traits<Allocator>::allocate(_alloc, _capacity);
+  }
+
+  void _construct(size_type n, size_type start_idx = 0) {
+    if (n > 0) {
+      for (size_type i = 0; i < n; i++) {
+        std::allocator_traits<Allocator>::construct(_alloc,
+                                                    _storage + start_idx + i);
+      }
+    }
+  }
+
+  void _construct(size_type n, const T &value, size_type start_idx = 0) {
+    if (n > 0) {
+      for (size_type i = 0; i < n; i++) {
+        std::allocator_traits<Allocator>::construct(
+            _alloc, _storage + start_idx + i, value);
+      }
+    }
+  }
+
+  template <typename Iter>
+  void _construct(Iter first, Iter last, size_type start_idx = 0) {
+    size_type num_eles = ::std::distance(first, last);
+    if (num_eles > 0) {
+      size_type i = 0;
+      for (auto iter = first; iter != last; iter++, i++) {
+        std::allocator_traits<Allocator>::construct(
+            _alloc, _storage + start_idx + i, *iter);
+      }
+    }
+  }
+
+  void _destroy(size_type n, size_type start_idx = 0) {
+    if (n > 0) {
+      for (size_type i = 0; i < n; i++) {
+        std::allocator_traits<Allocator>::destroy(_alloc,
+                                                  _storage + start_idx + i);
+      }
+    }
   }
 
 public:
   template <typename OtherA> operator ::std::vector<T, OtherA>() const {
     auto __tmp = ::std::vector<T, OtherA>(this->size());
     ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-              this->begin(), this->end(), __tmp.begin());
+                this->begin(), this->end(), __tmp.begin());
     return __tmp;
   }
-  device_vector()
-      : _alloc(get_default_queue()), _size(0), _capacity(_min_capacity()) {
+
+  device_vector(const Allocator &alloc)
```

Comment:
Can we give this a default argument, so we don't need the additional constructor in line 125?

Reply:
Yes, addressed in a few places.